### PR TITLE
[Ripple] Add better resizing animation support

### PR DESCRIPTION
**Pull request by yarneo**
As part of the effort for: #7268 to integrate Ripple into Chips, there were some improvements that were added into the Ripple component.

1. Chips support resizable chips on tap, which can be animated. To have the Ripple expand correctly based on the new animated bounds of the Chip, we added a synchronization between the animations. This is the same as what Ink currently offers as part of its implementation. This was tested along with PR #7270 to check that the animations were successfully synchronized.

This should land prior to #7270 to make sure that the Ripple integration with Chips has these improvements to ensure a smooth and correct integration.

Using the old Ink | Using Ripple without this improvement | Using Ripple with this improvement
------------ | ------------- | -------------
inkanimated | ripplebeforechange | afterchangeripple

### CODE REVIEW

**Comment by jverkoey** [comment]
Please add a gif of the resizing animation.

File: components/Ripple/src/MDCRippleView.m [hunk 1]
```diff
@@ -194,4 +206,42 @@ - (void)rippleLayerTouchUpAnimationDidEnd:(MDCRippleLayer *)rippleLayer {
...
+    // animation created by UIKit and copy the configuration to our custom animation.
+    CAAnimation *boundsAction = [self.animationSourceLayer animationForKey:@"bounds.size"];
+    if ([boundsAction isKindOfClass:[CABasicAnimation class]]) {
+      CABasicAnimation *animation = (CABasicAnimation *)[boundsAction copy];
+      animation.keyPath = self.keyPath;
+      animation.fromValue = self.fromValue;
+      animation.toValue = self.toValue;
+
+      [layer addAnimation:animation forKey:event];
+    }
```
Comment (jverkoey) [inline components/Ripple/src/MDCRippleView.m]: In cases where the animation is not a CABasicAnimation we'll fail to animate here; is that an acceptable outcome? Should we throw an assertion?

File: components/Ripple/src/MDCRippleView.m [hunk 2]
```diff
@@ -194,4 +206,42 @@ - (void)rippleLayerTouchUpAnimationDidEnd:(MDCRippleLayer *)rippleLayer {
...
+  }
+  return nil;
+}
+
+@end
+
+@implementation MDCRipplePendingAnimation
+
+- (void)runActionForKey:(NSString *)event object:(id)anObject arguments:(NSDictionary *)dict {
+  if ([anObject isKindOfClass:[CAShapeLayer class]]) {
```
Comment (jverkoey) [inline components/Ripple/src/MDCRippleView.m]: Use an early exit here (like a guard statement):

```objc
if (![anObject isKindOfClass:[CAShapeLayer class]]) {
  return;
}

CAShapeLayer *layer = (CAShapeLayer *)anObject;
```

File: components/Ripple/src/MDCStatefulRippleView.m [hunk 3]
```diff
@@ -171,7 +171,7 @@ - (void)setRippleHighlighted:(BOOL)rippleHighlighted {
     // If ripple becomes highlighted we initiate a ripple with animation.
     [self updateRippleColor];
     [self beginRippleTouchDownAtPoint:_lastTouch animated:_didReceiveTouch completion:nil];
-  } else {
+  } else if (!rippleHighlighted) {
```
Comment (jverkoey) [inline components/Ripple/src/MDCStatefulRippleView.m]: Is this change necessary for this PR? Can we break it out to a separate PR?

**Comment by yarneo** [comment]
I have added GIFs showing how Ink worked, how Ripple worked before this improvement, and how it works with this improvement.

**Comment by yarneo** [comment]
This improvement along with the Ripple integration, if Ripple is used, resolves #3056

File: components/Ripple/src/MDCStatefulRippleView.m (hunk 3 above)
Comment (yarneo) [reply components/Ripple/src/MDCStatefulRippleView.m]: This could be a separate PR, but is essential for the initial integration of Ripple with Chips. I can open a separate PR for this one liner. I just thought it wouldn't be a substantial burden to have both improvements (being small) in the same PR.

Comment (jverkoey) [reply components/Ripple/src/MDCStatefulRippleView.m]: Please break it out to a separate PR; it's not related to the rest of this PR and may risk getting rolled back if the other code above (which is inherently more complex) has to be reverted some day for any reason.

**Comment by jverkoey** [comment]
Screen Shot 2019-04-25 at 11 51 45 PM

Noticed one of the frames of the after animation looked a bit strange: is this just due to the artifacts of gif stippling, or is there a hard cut around the shadow?

File: components/Ripple/src/MDCRippleView.m (hunk 1 above)
Comment (yarneo) [reply components/Ripple/src/MDCRippleView.m]: yes, happy to throw an assertion here. This mainly syncs to this chip resize animation: https://github.com/material-components/material-components-ios/blob/develop/components/Chips/src/MDCChipCollectionViewCell.m#L51

File: components/Ripple/src/MDCRippleView.m (hunk 2 above)
Comment (yarneo) [reply components/Ripple/src/MDCRippleView.m]: Done.

File: components/Ripple/src/MDCStatefulRippleView.m (hunk 3 above)
Comment (yarneo) [reply components/Ripple/src/MDCStatefulRippleView.m]: Removed

**Comment by yarneo** [comment]
The shadow hard cut seems like a GIF rendering thing and doesn't occur on the component, but good catch finding this.

**Review by jverkoey — APPROVED** [review]
